File: go/139.go
//
//
//package main
//
//import "fmt"

// 6 star, 动态规划，dp[i]代表s[:i]能否被数组里的字符串组合成，如果s[j:i]在数组中，且dp[j]为true，则dp[i]为true
func wordBreak(s string, wordDict []string) bool {
	length := len(s)
	dp := make([]bool, length+1)
	dict := make(map[string]string, len(wordDict))
	for _, word := range wordDict {
		dict[word] = ""
	}
	dp[0] = true
	for i := 1; i < length+1; i++ {
		for j:=0; j < i; j++ {
			_, ok := dict[s[j:i]]
			if dp[j] && ok {
				dp[i] = true
			}
		}
	}
	return dp[length]

}


//func main() {
//	fmt.Println(wordBreak("leetcode", []string{"leet", "code"}))
//}

File: v2/139.py




class Solution:
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        # 6 star, 动态规划，必须掌握
        # dp[i]表示s[:i]字符串是否能拆分成符合要求的字符串，如果存在s[j:i]是在字符列表里以及dp[j]为真，则dp[i]为真
        str_set = set(wordDict)
        length = len(s)
        dp = [False for _ in range(length+1)]
        dp[0] = 1
        for i in range(1, length+1):
            for j in range(i):
                if dp[j] and s[j:i] in str_set:
                    dp[i] = True
        return dp[-1]
